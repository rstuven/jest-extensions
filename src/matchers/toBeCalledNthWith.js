// Based on jest/packages/jest-matchers/src/spyMatchers.js

const isSpy = spy => spy.calls && typeof spy.calls.count === 'function'
const {equals} = require('jest-matchers/build/jasmine-utils')
const {
  // ensureExpectedIsNumber,
  // ensureNoExpected,
  // EXPECTED_COLOR,
  matcherHint,
  pluralize,
  printExpected,
  printReceived,
  // printWithType,
  RECEIVED_COLOR,
} = require('jest-matcher-utils')
const RECEIVED_NAME = {
  'mock function': 'jest.fn()',
  spy: 'spy',
}
const CALL_PRINT_LIMIT = 3
const LAST_CALL_PRINT_LIMIT = 1

const formatReceivedCallIndex = (calls, index) => {
  if (calls.length) {
    const printedCall = printReceived(calls[index])
    return `But it was called with:\n  ` + printedCall
  } else {
    return `But it was ${RECEIVED_COLOR('not called')}.`
  }
}

const ensureMock = (mockOrSpy, matcherName) => {
  if (
    !mockOrSpy ||
    (mockOrSpy.calls === undefined || mockOrSpy.calls.all === undefined) &&
    mockOrSpy._isMockFunction !== true
  ) {
    throw new Error(
      matcherHint('[.not]' + matcherName, 'jest.fn()', '') + '\n\n' +
      `${RECEIVED_COLOR('jest.fn()')} value must be a mock function or spy.\n` +
      printWithType('Received', mockOrSpy, printReceived),
    )
  }
}

module.exports = (expect) => {
  expect.extend({
    toBeCalledNthWith(received: any, index: number, ...expected: any) {
      const matcherName = 'toBeCalledNthWith'
      ensureMock(received, matcherName)

      const receivedIsSpy = isSpy(received)
      const type = receivedIsSpy ? 'spy' : 'mock function'
      const calls = receivedIsSpy
        ? received.calls.all().map(x => x.args)
        : received.mock.calls
      const pass = equals(calls[index], expected)

      const message = pass
        ? () => matcherHint('.not' + matcherName, RECEIVED_NAME[type]) + '\n\n' +
          `Expected ${type} to not have been called with:\n` +
          `  ${printExpected(expected)}`
        : () => matcherHint(matcherName, RECEIVED_NAME[type]) + '\n\n' +
          `Expected ${type} to have been called with:\n` +
          `  ${printExpected(expected)}\n` +
          formatReceivedCallIndex(calls, index)

      return { message, pass }
    },
  })
}
