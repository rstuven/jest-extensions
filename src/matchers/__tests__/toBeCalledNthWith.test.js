
describe('toBeCalledNthWith', () => {
  let matchersObject

  beforeEach(() => {
    jest.resetAllMocks()
    const expectObject = {
      extend(mo) { matchersObject = mo }
    }
    require('../toBeCalledNthWith')(expectObject)
  })

  it('pass: true', () => {
    const mock = jest.fn(function name() {})
    mock('call1', 'callA')
    mock('call2', 'callB')
    const result1 = matchersObject.toBeCalledNthWith(mock, 0, 'call1', 'callA')
    const result2 = matchersObject.toBeCalledNthWith(mock, 1, 'call2', 'callB')
    expect(result1.pass).toBe(true)
    expect(result2.pass).toBe(true)
  })

  it('pass: false', () => {
    const mock = jest.fn(function name() {})
    mock('call1', 'callA')
    mock('call2', 'callB')
    const result = matchersObject.toBeCalledNthWith(mock, 1, 'call0', 'callX')
    expect(result.pass).toBe(false)
    expect(result.message()).toMatchSnapshot()
  })

})
