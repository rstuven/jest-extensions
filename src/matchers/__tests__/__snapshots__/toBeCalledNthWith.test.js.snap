exports[`toBeCalledNthWith pass: false 1`] = `
"[2mexpect([22m[31mjest.fn()[39m[2m)toBeCalledNthWith([22m[32mexpected[39m[2m)[22m

Expected mock function to have been called with:
  [32m[\"call0\", \"callX\"][39m
But it was called with:
  [31m[\"call2\", \"callB\"][39m"
`;
